Improve ResourceIndexer tests

epicadk opened the pull request:
Fixes: #282

**Description:**
Clear and concise code change description. 

1.  Renamed unit tests to signify the code path they test. 
2.  Use minimally constructed resources to test specific types
3. Add integration tests that use resources parsed from JSON files and verify that all indices are created.
4. Removed setup method so that only the resource being used in the test is parsed as compared to all resources when using the setup method. 

**Type** 
  Testing


**Screenshots (if applicable)**

jingtang10 commented:
is there any update on this one? thanks!

jingtang10 commented:
please make sure @deepankarb also has a chance to review once done. thanks!

epicadk commented:
> is there any update on this one? thanks!

Sorry for the delay. Will be done within a day.

epicadk commented:
Sorry something urgent came up. I've completed adding unit tests. Will finish integration tests soon.

epicadk commented:
I need to place the integration tests in a different folder. I was thinking I can create a separate directory for unit tests and integration tests under the test directory. Any suggestions?

epicadk commented:
I've also added some tests in #332 for the date types. I've covered all cases except the integer case of the number index. I can't seem to find a resource that uses a search parameter of that type.  Even the example given on the specification website doesn't work.

jingtang10 commented on core/src/test/java/com/google/android/fhir/ResourceIndexerTest.kt:
```diff
@@ -0,0 +1,280 @@
+/*
```
I don't think it's right to create another file in this package. This should be in the same directory as the other file.

epicadk replied:
> I don't think it's right to create another file in this package. This should be in the same directory as the other file.

Reverted this change

jingtang10 reviewed (changes requested):
It might be ok to keep everything in the same file, but you should have two classes of test cases

1. cases that test specific code paths e.g. `string_empty_shouldIndexNothing` or `int_shouldCreateIndex` etc etc. We would use minimally constructed resources to test specific types.
2. cases where we use more complex resources and assert that all indecies that should be created are created.

I think these two classes of tests can probably co-exist in the same file if named sensibly.

epicadk commented:
I've kept the order similar as they occur in the resource indexer. I've also included tests for null and empty values wherever possible.

jingtang10 commented on engine/src/test/java/com/google/android/fhir/index/ResourceIndexerTest.kt:
```diff
@@ -26,548 +26,803 @@ import com.google.android.fhir.index.entities.ReferenceIndex
+    testLocation = testingUtils.readFromFile(Location::class.java, "/location-example-hl7hq.json")
   }
 
+  /** Integration tests for ResourceIndexer. */
```
can you move the 'integration' tests after the 'unit' tests?

jingtang10 commented on engine/src/test/java/com/google/android/fhir/index/ResourceIndexerTest.kt:
```diff
@@ -26,548 +26,803 @@ import com.google.android.fhir.index.entities.ReferenceIndex
-        resourceIndices.stringIndices.any { stringIndex ->
-          stringIndex.path.equals("Patient.name.given")
-        }
+  fun index_Patient_shouldIndexLastUpdatedDate() {
```
the point of these unit test cases is to cover all code paths so i think they need to be re-organized by index type, and please rename them, e.g.:

```
fun index_string_null
fun index_string
...

fun date_period_someCondition
fun date_period_someOtherCondition
...
```

jingtang10 reviewed (changes requested):
A couple of high-level comments. Please address! Thanks!

epicadk commented:
Noticed a bug in the date indexer while testing so I made the corresponding change. The value to should be inclusive (makes querying easier using the SQL between operator.)

jingtang10 commented:
@epicadk did you remove all the other changes in the test class you made?

epicadk commented:
> @epicadk did you remove all the other changes in the test class you made?

Yup all of the tests are in the same file and I've renamed the tests as requested [here](https://github.com/google/android-fhir/pull/387#discussion_r616014331). I've also updated the pr description to explain the changes I have made.

jingtang10 commented on engine/src/test/java/com/google/android/fhir/index/ResourceIndexerTest.kt:
```diff
@@ -26,542 +26,831 @@ import com.google.android.fhir.index.entities.ReferenceIndex
+    assertThat(resourceIndices.uriIndices).isEmpty()
+
+    assertThat(resourceIndices.dateIndices)
       .contains(
```
```suggestion
      .containsExactly(
```

jingtang10 reviewed (changes requested):
@epicadk Really happy with this - this is a very good PR! Thank you so much for this!

Apologies for the delay in reviewing. Please let me know once you're done with the few small comments I made.

epicadk commented:
@jingtang10 divided the tests in 3 parts.

1.  The creation of resources 
2. Actual code for indexing
3. assertions

jingtang10 reviewed (approved):
Great! Thanks!
